third_party/amd/lib/TritonAMDGPUToLLVM/MemoryOpToLLVM.cpp

```diff
@@ -44,8 +39,327 @@ class TransLocalLoadOpConversion
+    unsigned otherLanes = 1;
+    if (isaFamily == AMD::ISAFamily::CDNA4) {
+      otherLanes = (bitWidth == 8) ? 2 : 4;
+    } else if (doubleB8Contiguity) {
+      otherLanes = 2;
+    }
+
+    if (doubleB8Contiguity) {
+      fullTile =
+          tile *
+          LinearLayout::identity1D(ldsParams.needContigReg / 2, kReg, kAddr) *
+          LinearLayout::identity1D(otherLanes, kLane, kAddr) *
+          LinearLayout::identity1D(2, kReg, kAddr) *
+          LinearLayout::identity1D(missingLanes / otherLanes, kLane, kAddr);
+    } else {
+      fullTile =
+          tile * LinearLayout::identity1D(otherLanes, kLane, kAddr) *
+          LinearLayout::identity1D(ldsParams.needContigReg, kReg, kAddr) *
+          LinearLayout::identity1D(missingLanes / otherLanes, kLane, kAddr);
+    }
+
+    if (cvt.getInDimSize(kReg) < fullTile.getInDimSize(kReg)) {
+      return failure();
+    }
+
+    // Find if there is a register permutation that allows us to divideLeft
+    // Note: our tile doesn't currently need register permutations (reg size 1)
+    ColumnAction permDivide;
+    if (auto maybePermutation = regPermForDivide(cvt, tile, /*left=*/true)) {
+      permDivide = maybePermutation.value();
+    } else {
+      return failure();
+    }
+    cvt = permDivide.apply(cvt);
+    accPermReg = accPermReg.leftCompose(permDivide);
+    auto maybeQuot = divideLeft(cvt, tile);
+    if (!maybeQuot.has_value()) {
+      return failure();
+    }
+
+    // From here on we perform the lowering
+    auto reps = zerosLike(tile) * maybeQuot.value();
+
+    // Sanity check
+    assert(fullTile.getInDimSize(kReg) * bitWidth == ldsParams.instBitWidth);
+
+    // If we are lowering a subslice, the subslice offsets shall not touch the
+    // contiguous part of the tile
+    if (maskSpanAffineOffset & (tile.getOutDimSizeLog2(kOffset) - 1)) {
+      return failure();
+    }
+
+    // Choose the vectorisation factor.
+    // We don't need to vectorize further so set vec to 1.
+    // Keep fullTileVec to match ldmatrix lowering.
+    auto vec = 1;
+    auto fullTileVec = fullTile * LinearLayout::identity1D(vec, kReg, kAddr);
+
+    // just add warps as compose belowe requires the dimensions of both layouts
+    // to agree
+    fullTileVec *= LinearLayout::identity1D(1, kWarp, kAddr);
+    // fullTile.invert() is a map from kOffset, kAddr into kReg, kLane, kWarp
+    // addrToOffset gives us a map from kAddr into kOffset, which is the map of
+    // the addresses each lane should hold
+    auto addrToOffset = fullTileVec.invert().compose(reps);
+    // sanity check
+    assert(addrToOffset.getInDimSizeLog2(kAddr) >= 3 &&
+           addrToOffset.getInDimSizeLog2(kAddr) <= 6);
+
+    LinearLayout addrLayout =
+        LinearLayout({{kLane, addrToOffset.getBases().lookup(kAddr)},
+                      {kWarp, reps.getBases().lookup(kWarp)}},
+                     {{kOffset, reps.getOutDimSize(kOffset)}}, false);
```

Comment:
Does this instruction also require lane i to hold the offset of the row i? If that is the case, this would imply that this instruction would transpose a matrix with between 2^3 to 2^6 rows. Is that the case?